File: src/main/scala/com/yogpc/qp/packet/advpump/AdvPumpStatusMessage.java
package com.yogpc.qp.packet.advpump;

import java.util.function.Supplier;

import com.yogpc.qp.machines.advpump.TileAdvPump;
import com.yogpc.qp.packet.IMessage;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.network.PacketBuffer;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.fml.network.NetworkEvent;

/**
 * To Client only.
 */
public class AdvPumpStatusMessage implements IMessage<AdvPumpStatusMessage> {

    ResourceLocation dim;
    BlockPos pos;
    boolean placeFrame;
    CompoundNBT nbtTagCompound;

    public static AdvPumpStatusMessage create(TileAdvPump pump) {
        AdvPumpStatusMessage message = new AdvPumpStatusMessage();
        message.dim = IMessage.getDimId(pump.getWorld());
        message.pos = pump.getPos();
        message.nbtTagCompound = pump.getUpdateTag();
        message.placeFrame = pump.placeFrame();
        return message;
    }

    @Override
    public AdvPumpStatusMessage readFromBuffer(PacketBuffer buffer) {
        pos = buffer.readBlockPos();
        nbtTagCompound = buffer.readCompoundTag();
        placeFrame = buffer.readBoolean();
        dim = buffer.readResourceLocation()();
        return this;
    }

    @Override
    public void writeToBuffer(PacketBuffer buffer) {
        buffer.writeBlockPos(pos).writeCompoundTag(nbtTagCompound).writeBoolean(placeFrame).writeInt(dim);
    }

    @Override
    public void onReceive(Supplier<NetworkEvent.Context> ctx) {
        IMessage.findTile(ctx, pos, dim, TileAdvPump.class)
            .ifPresent(pump -> ctx.get().enqueueWork(() -> pump.receiveStatusMessage(placeFrame, nbtTagCompound)));
    }
}

File: src/main/scala/com/yogpc/qp/packet/advpump/AdvPumpStatusMessage.scala
package com.yogpc.qp.packet.advpump

import java.io.IOException

import com.yogpc.qp.QuarryPlus
import com.yogpc.qp.packet.IMessage
import com.yogpc.qp.tile.TileAdvPump
import net.minecraft.client.Minecraft
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.PacketBuffer
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext
import net.minecraftforge.fml.relauncher.{Side, SideOnly}

/**
  * To Client only.
  */
object AdvPumpStatusMessage {
    def create(pump: TileAdvPump): AdvPumpStatusMessage = new AdvPumpStatusMessage(pump)
}

class AdvPumpStatusMessage(pump: TileAdvPump) extends IMessage {
    private[this] var dim = 0
    private[this] var pos = BlockPos.ORIGIN
    private[this] var placeFrame = false
    private[this] var nbtTagCompound: NBTTagCompound = _

    def this() {
        this(null)
    }

    if (pump != null) {
        dim = pump.getWorld.provider.getDimension
        pos = pump.getPos
        nbtTagCompound = pump.getUpdateTag
        placeFrame = pump.placeFrame
    }

    @throws[IOException]
    override def fromBytes(buffer: PacketBuffer): Unit = {
        pos = buffer.readBlockPos
        nbtTagCompound = buffer.readCompoundTag
        placeFrame = buffer.readBoolean
        dim = buffer.readInt
    }

    override def toBytes(buffer: PacketBuffer): Unit = buffer.writeBlockPos(pos).writeCompoundTag(nbtTagCompound).writeBoolean(placeFrame).writeInt(dim)

    @SideOnly(Side.CLIENT)
    override def onRecieve(message: IMessage, ctx: MessageContext): IMessage = {
        val world = QuarryPlus.proxy.getPacketWorld(ctx.netHandler)
        if (world.provider.getDimension == dim) {
            world.getTileEntity(pos) match {
                case pump: TileAdvPump => Minecraft.getMinecraft.addScheduledTask(IMessage.toRunnable(() =>
                    pump.recieveStatusMessage(placeFrame, nbtTagCompound)
                ))
                case _ =>
            }
        }
        null
    }
}